# TP2 - HMLA408: Joseph Salmon

## Question: importer la fonction chisquare

In [1]:
from scipy.stats import chisquare, norm, shapiro, poisson, chi2, chi2_contingency
import numpy as np

chisquare?

## Question: cobayes

In [2]:
cobaye = np.array([33, 13, 15, 3])

In [3]:
mendel = np.array([9/16, 3/16, 3/16, 1/16])

In [4]:
chisq, pval  = chisquare(cobaye, f_exp=mendel * np.sum(cobaye))
print(chisq, pval)

1.3333333333333333 0.7212333746277604


## Conclusion du test:
la pvalue est de 0.72 ce qui est très grand, on ne rejette pas au niveau 0.05% par exemple

In [5]:
from scipy.stats import chi2
print(1-chi2.cdf(chisq, 3), pval)
# (np.testing.assert_almost_equal(1-chi2.cdf(a, 3), b) # exemple testing that two quantities are numerically equal

0.7212333746277604 0.7212333746277604


1-chi2.cdf(chisq, 3) est la probabilité qu'une variable aléatoire qui suit une loi de chi2 à 3 degré de liberté ait une valeur supérieure à chisq=1.333333. Cette probabilité est égale à la p-value.


## Question: importer pandas

In [6]:
import pandas as pd

In [7]:
echantillon_gaussien = np.random.randn(50)
bins = np.array([-np.inf, -2, -1, 0, 1, 2, np.inf]) #ATTENTION sans np.array() l'ordre est différent
print(echantillon_gaussien)

[-1.31811496  0.07793067  0.30279908 -0.23579763  0.03618076  1.51261777
 -1.93531333  0.51220733  0.27475523  0.46368898  0.39030049  1.22131413
 -1.19451943 -2.01329058 -0.11337173  0.03793557  1.19058797 -0.2660515
  0.65924948 -1.5960921  -0.16341216 -1.03217257 -0.72481432 -0.29028289
  0.17347548  1.44054729 -0.24000309  1.10054452  0.62995592  0.15114704
 -0.92233237  0.27610918 -0.28982219  1.65027539 -1.24236028 -0.02382412
  0.38431724 -0.73529955  0.7910851  -1.90004272 -0.08843393 -0.7517388
 -0.58259617 -0.0229063   0.31920862  0.15552755  0.88868584  0.77739361
 -0.5592531   0.67912283]


In [8]:
echantillon_regroupe = pd.cut(echantillon_gaussien,bins=bins)  # jamais d'accent dans les noms de variables
echantillon_regroupe[22]

Interval(-1.0, 0.0, closed='right')

In [9]:
echantillon_regroupe.value_counts()

(-inf, -2.0]     1
(-2.0, -1.0]     7
(-1.0, 0.0]     16
(0.0, 1.0]      20
(1.0, 2.0]       6
(2.0, inf]       0
dtype: int64

In [10]:
pd.value_counts(echantillon_regroupe) # weird behaviour : ordering chaning w.r.t. to .value_counts()

(0.0, 1.0]      20
(-1.0, 0.0]     16
(-2.0, -1.0]     7
(1.0, 2.0]       6
(-inf, -2.0]     1
(2.0, inf]       0
dtype: int64

In [11]:
effectifs_classes = echantillon_regroupe.value_counts()

## Question: fonction de répartition


In [12]:
norm.cdf(0) #cdf : cumulative distribution function = fonction de répartition, ici de la loi normale centrée réduite

0.5

In [13]:
norm.cdf(1.96)

0.9750021048517795

In [14]:
p1 = norm.cdf(bins[1])
#calcul de la probabilité que X soit inférieur à -2 (X normale centrée réduite)
p2 = norm.cdf(bins[2]) - norm.cdf(bins[1])
#calcul de la probabilité que X soit compris entre -2 (bins[1]) et -1 (bins[2])
p3 = norm.cdf(bins[3]) - norm.cdf(bins[2])
p4 = norm.cdf(bins[4]) - norm.cdf(bins[3])
p5 = norm.cdf(bins[5]) - norm.cdf(bins[4])
p6 = 1 - norm.cdf(bins[5])
probas = np.array([p1, p2, p3, p4, p5, p6])
print(probas)
#on calcule les probabilités théoriques

[0.02275013 0.13590512 0.34134475 0.34134475 0.13590512 0.02275013]


La commande suivante fait exactement la même chose que la précédente (mais serait plus facile à mettre en oeuvre avec beaucoup de boîtes)

In [15]:
np.diff(norm.cdf((bins)))

array([0.02275013, 0.13590512, 0.34134475, 0.34134475, 0.13590512,
       0.02275013])

## Question: test du $\chi^2$

In [16]:
np.diff(norm.cdf(bins)) * echantillon_regroupe.shape[0]

array([ 1.1375066,  6.7952561, 17.0672373, 17.0672373,  6.7952561,
        1.1375066])

In [17]:
chisquare(effectifs_classes, np.diff(norm.cdf(bins)) * echantillon_regroupe.shape[0])


Power_divergenceResult(statistic=1.8240571305720577, pvalue=0.8729148516869414)


* `effectifs_classes` = effectifs observés dans l'échantillon.
* `np.diff(norm.cdf(bins)) * echantillon_regroupe.shape[0]` = effectifs attendus


## Conclusion:
On ne rejette pas l'hypothèse aux seuil 1, 5 10 %...pvalue=0.14

# Question: reprise du test avec d'autres intervalles (plus larges effectifs)

In [18]:
echantillon_gaussien = np.random.randn(100)
bins = np.array([-np.inf, -1.5, -0.5, 0.5, 1.5, np.inf])
# jamais d'accent dans les noms de variables
echantillon_regroupe = pd.cut(echantillon_gaussien, bins=bins)
effectifs_classes = echantillon_regroupe.value_counts()
chisquare(effectifs_classes, np.diff(norm.cdf(bins))
          * echantillon_regroupe.shape[0])

Power_divergenceResult(statistic=8.625419898212089, pvalue=0.07117540925837786)

# Question: Shapiro-Wilk
La loi de la statistique de test sous H0: "la distribution de l'échantillon est gaussienne"
est calculable par Python et on obtient directement la statistique associée et la p-value.

In [19]:
shapiro(echantillon_gaussien)
# ici par exemple p-value 0.68, donc on ne rejette pas H0 (ce qui est attendu vu que nos données sont bien Gaussiennes) 

(0.9866547584533691, 0.41482946276664734)

In [20]:
echantillon_exponential = np.random.exponential(1,100)
shapiro(echantillon_exponential)
# Ici rejet clair!

(0.7444982528686523, 6.7151722550617965e-12)

## Question: Download

In [21]:
from download import download
url = "http://josephsalmon.eu/enseignement/datasets/hcmv.data"
# url = "http://www.stat.berkeley.edu/users/statlabs/data/hcmv.data" # backup url, without header.
path_target = "./hcmv.data"
download(url, path_target, replace=False)

Replace is False and data exists, so doing nothing. Use replace==True to re-download the data.


'./hcmv.data'

In [22]:
df_hmcv = pd.read_csv("hcmv.data", sep='\s+') # \s+ : for hanlding spaces
df_hmcv.head(n=10)  # df stands for Data Frame

location
0       177
1      1321
2      1433
3      1477
4      3248
5      3255
6      3286
7      7263
8      9023
9      9084

## Question lambda_hat

In [23]:
n_basis = 229354
n_palindromes = df_hmcv.count()['location']
lambda_hat = n_palindromes / n_basis * 4000
print(lambda_hat)
n_palindromes

5.162325488110083


296

In [24]:
new_bins = np.concatenate([[-np.inf],np.arange(1, n_basis, step=4000),[np.inf]])

In [25]:
counts_bin = np.array([-np.inf, 2, 3, 4, 5, 6, 7, 8, np.inf])
# jamais d'accent dans les noms de variables
hmcv_regroupe = pd.cut(df_hmcv['location'], bins=new_bins)
counts_palindrome = hmcv_regroupe.value_counts(sort=False)
count_regroupe = pd.cut(counts_palindrome, counts_bin,
                        right=True).value_counts(sort=False)

In [26]:
count_regroupe

(-inf, 2.0]     9
(2.0, 3.0]      8
(3.0, 4.0]     10
(4.0, 5.0]      9
(5.0, 6.0]      8
(6.0, 7.0]      5
(7.0, 8.0]      4
(8.0, inf]      6
Name: location, dtype: int64

In [27]:
bins_poisson = [-np.inf, 2, 3, 4, 5, 6, 7, 8, np.inf]
count_regroupe = pd.cut(counts_palindrome, bins_poisson, right=True, labels=[
                             '0-2', '3', '4', '5', '6', '7', '8','9-']).value_counts(sort=False)

In [28]:
count_regroupe

0-2     9
3       8
4      10
5       9
6       8
7       5
8       4
9-      6
Name: location, dtype: int64

In [29]:
np.diff(poisson.cdf(bins_poisson, lambda_hat))

array([0.11162934, 0.13134557, 0.16951215, 0.17501537, 0.15058105,
       0.11104977, 0.07165938, 0.07920736])

In [30]:
poisson.cdf(bins_poisson, lambda_hat)

array([0.        , 0.11162934, 0.24297491, 0.41248705, 0.58750243,
       0.73808348, 0.84913325, 0.92079264, 1.        ])

In [31]:
np.diff(poisson.cdf(bins_poisson, lambda_hat))

array([0.11162934, 0.13134557, 0.16951215, 0.17501537, 0.15058105,
       0.11104977, 0.07165938, 0.07920736])

In [32]:
count_theoric = np.diff(poisson.cdf(bins_poisson, lambda_hat)) * count_regroupe.sum()
chi2_stat, chi2_pvalue = chisquare(count_regroupe, count_theoric, ddof=1)
# Attention au nombre de degrés de libertés, ddof=1 ici car on ne connait pas le parametre lambda
print(chi2_stat, chi2_pvalue)

1.907642554439036 0.9279935984980819


# Question: $\chi^2$ à la main

In [33]:
pvalue_manual = 1-chi2.cdf(1.907642554439036,df=6)
print(pvalue_manual)

0.9279935984980819


In [34]:
np.testing.assert_almost_equal(pvalue_manual, chi2_pvalue) 

# Chargement de la base de données 

In [35]:
url = "http://josephsalmon.eu/enseignement/datasets/babies23.data"
path_target = "./babies23.data"
download(url, path_target, replace=False)

Replace is False and data exists, so doing nothing. Use replace==True to re-download the data.


'./babies23.data'

# Preoprocessing: only run once or big trouble (think about it!)

In [36]:
is_preprocessing_done = 0  # init at 0,if greater don't redo it
pd.options.display.max_rows = 8  # set not to display to many lines in pandas
pd.set_option('precision', 0)  # set to display number at precision 0 in pandas

# Lecture de la base de données et constructions d'un dataframe

In [37]:
df_babies = pd.read_csv("babies23.data", skiprows=38, sep='\s+') # \s+ : for hanlding spaces
df_babies.head(n=10)  # df stands for Data Frame

id  pluralty  outcome  date  gestation  sex   wt  parity  race  age  ...  \
0    15         5        1  1411        284    1  120       1     8   27  ...   
1    20         5        1  1499        282    1  113       2     0   33  ...   
2    58         5        1  1576        279    1  128       1     0   28  ...   
3    61         5        1  1504        999    1  123       2     0   36  ...   
..  ...       ...      ...   ...        ...  ...  ...     ...   ...  ...  ...   
6   102         5        1  1449        244    1  138       4     7   33  ...   
7   129         5        1  1562        245    1  132       2     7   23  ...   
8   142         5        1  1408        289    1  120       3     0   25  ...   
9   148         5        1  1568        299    1  143       3     0   30  ...   

    drace  dage  ded  dht  dwt  marital  inc  smoke  time  number  
0       8    31    5   65  110        1    1      0     0       0  
1       0    38    5   70  148        1    4      0     0       0  
2       5    32    1   99  999        1    2      1     1       1  
3       3    43    4   68  197        1    8      3     5       5  
..    ...   ...  ...  ...  ...      ...  ...    ...   ...     ...  
6       7    37    4   99  999        1   98      0     0       0  
7       7    23    4   71  192        1    2      0     0       0  
8       3    26    1   70  180        0    2      0     0       0  
9       0    34    5   99  999        1    2      1     1       4  

[10 rows x 23 columns]

In [38]:
cont_table = pd.crosstab(df_babies['ed'], df_babies['smoke'])
cont_table

smoke    0    1   2   3  9
ed                        
0       12    4   1   2  0
1       57  102  12  10  2
2      202  176  40  22  4
3       21   33   3   6  2
4      138  102  21  35  2
5      109   65  18  27  0
7        5    1   0   1  0
9        0    1   0   0  0

In [39]:
chi2, p_value, dof, ex = chi2_contingency(cont_table)

In [40]:
p_value
# on rejette donc l'hypothèse H_0 d'indépendance entre le niveau d'éducation et le fait de fumer sur ces données.
# on accepte donc l'hyphèse quelle sont liées...

6.000912536921962e-05

In [41]:
chi2

66.30862239116085

In [42]:
dof

28

# Question: comparer à la main avec un degré de liberté bien choisi

In [43]:
ex 

array([[8.36245955e+00, 7.44012945e+00, 1.46035599e+00, 1.58333333e+00,
        1.53721683e-01],
       [8.05436893e+01, 7.16601942e+01, 1.40655340e+01, 1.52500000e+01,
        1.48058252e+00],
       [1.95417476e+02, 1.73864078e+02, 3.41262136e+01, 3.70000000e+01,
        3.59223301e+00],
       [2.86084142e+01, 2.54530744e+01, 4.99595469e+00, 5.41666667e+00,
        5.25889968e-01],
       [1.31158576e+02, 1.16692557e+02, 2.29045307e+01, 2.48333333e+01,
        2.41100324e+00],
       [9.63883495e+01, 8.57572816e+01, 1.68325243e+01, 1.82500000e+01,
        1.77184466e+00],
       [3.08090615e+00, 2.74110032e+00, 5.38025890e-01, 5.83333333e-01,
        5.66343042e-02],
       [4.40129450e-01, 3.91585761e-01, 7.68608414e-02, 8.33333333e-02,
        8.09061489e-03]])

In [44]:
chisquare(cont_table.values.ravel(), f_exp=ex.ravel(), ddof=cont_table.size - 1 - dof)

Power_divergenceResult(statistic=66.30862239116085, pvalue=6.000912536921962e-05)

Remarque: il se trouve que la loi du $\chi^2$ pour ce type de test est une distribution du $\chi^2_{(c-1)(l-1)}$
avec $c$ le nombre de colonnes, et $l$ le nombre de ligne du tableau.

## Courbes ROC

In [45]:
"""
Graphics for Roc curve
author: portier, salmon
"""

import numpy as np
import matplotlib.pyplot as plt
import os
from matplotlib import rc
import seaborn as sns
from scipy.stats import norm
import matplotlib.patches as mpatches

###############################################################################
# os.chdir('/Users/francois/Desktop/Cours/mdi720/cours/5-IC_Tests/')
cwd = os.getcwd()

dirname = "../prebuiltimages/"
if not os.path.exists(dirname):
    os.mkdir(dirname)
imageformat = '.pdf'


def my_saving_display(fig, dirname, filename, imageformat):
    """"saving faster"""
    dirname + filename + imageformat
    image_name = dirname + filename + imageformat
    fig.savefig(image_name)


###############################################################################
# new color

rc('font', **{'family': 'sans-serif', 'sans-serif': ['Computer Modern Roman']})
params = {'axes.labelsize': 12,
          'font.size': 10,
          'legend.fontsize': 8,
          'xtick.labelsize': 0,
          'ytick.labelsize': 0,
          'text.usetex': True,
          'figure.figsize': (8, 6)}
plt.rcParams.update(params)

sns.set_context("poster")
sns.set_style("white")
sns.set_palette("colorblind")
sns.despine(ax=None, top=True, right=True, left=True, bottom=True)
color_blind_list = sns.color_palette("colorblind", 8)
my_orange = color_blind_list[2]
my_green = color_blind_list[1]
my_blue = color_blind_list[0]


###############################################################################
# Gaussian pdf

xs = np.linspace(-1, 9, num=150)
bbox_props = dict(boxstyle="round", fc="w", ec="0.5", alpha=0.9)

In [46]:
def plot_roc_n_graph(q=1.75, mu_normal=1.5, sigma2_normal=1, mu_abnormal=3.2, sigma2_abnormal=1.2):
    y1 = norm.pdf(xs, loc=mu_normal, scale=np.sqrt(sigma2_normal))
    y2 = norm.pdf(xs, loc=mu_abnormal, scale=np.sqrt(sigma2_abnormal))

    fig, ax = plt.subplots(1, 2, figsize=(13, 6))
    FN = 1 - norm.cdf(q, loc=mu_normal, scale=np.sqrt(sigma2_normal))
    FP = 1 - norm.cdf(q, loc=mu_abnormal, scale=np.sqrt(sigma2_abnormal))

    FN_full = 1 - norm.cdf(xs, loc=mu_normal, scale=np.sqrt(sigma2_normal))
    FP_full = 1 - norm.cdf(xs, loc=mu_abnormal, scale=np.sqrt(sigma2_abnormal))

    ax[0].plot(FN_full, FP_full, 'k', linewidth=7)
    ax[0].plot(FN, FP, 'o', c='red', markersize=20)
    ax[0].set_xlabel('FP', fontsize=30)
    ax[0].set_ylabel('1-FN', fontsize=30)
    ax[0].tick_params(axis='both', which='major', labelsize=30)
    ax[0].set_xlim([-0.1, 1.3])
    ax[0].set_ylim([-0.1, 1.3])

    mod_norm = 1. / np.sqrt(2 * 3.14 * sigma2_normal)
    mod_abnorm = 1. / np.sqrt(2 * 3.14 * sigma2_abnormal)

    ax[1].text(mu_normal - 1.9, mod_norm - .21, '$H_0$', ha="center",
               va="bottom", size=20, bbox=bbox_props)
    ax[1].text(mu_abnormal + 1.49, mod_abnorm - 0.1, '$H_1$', ha="center",
               va="bottom", size=20, bbox=bbox_props)
    ax[1].text(q + .2, np.max([y1, y2]) + 0.05, '$q= ' + str(q) + '$',
               fontsize=30, fontweight='bold')

    ax[1].plot(xs, y1, color='k', linewidth=1)
    ax[1].plot(xs, y2, color='k', linewidth=1)

    ax[1].fill_between(xs, y2, where=xs <= q, facecolor=color_blind_list[1])
    ax[1].fill_between(xs, y1, where=xs >= q, facecolor=color_blind_list[0])

    ax[1].set_xlim([xs.min(), xs.max()])
    ax[1].set_ylim([-0.05, np.max([y1, y2]) + .1])
    ax[1].axvline(x=q, color='k', linewidth=5)
    sns.set_style("ticks")
    sns.despine()
    classes = ['FN', 'FP']
    class_colours = [color_blind_list[1], color_blind_list[0]]
    recs = []
    for i in range(0, len(class_colours)):
        recs.append(mpatches.Rectangle((0, 0), 1, 1, fc=class_colours[i]))
    leg = plt.legend(recs, classes, loc=1, fontsize=14)

    filename = 'fig_'
    plt.tight_layout()
    plt.show()
#     my_saving_display(fig, dirname, filename, imageformat)

In [47]:
from ipywidgets import interact

interact(plot_roc_n_graph,
         mu_normal= (1,3,0.5),
         sigma2_normal = (0.2, 2, 0.2),
         mu_abnormal= (1, 5, 0.5),
         sigma2_abnormal = (0.2, 2, 0.2),
         q=(0, 8, 0.05))

interactive(children=(FloatSlider(value=1.75, description='q', max=8.0, step=0.05), FloatSlider(value=1.5, des…

<function __main__.plot_roc_n_graph(q=1.75, mu_normal=1.5, sigma2_normal=1, mu_abnormal=3.2, sigma2_abnormal=1.2)>